### PR TITLE
CLI improvements

### PR DESCRIPTION
**Pull request by dkpoult**
This PR adds some improvements to how the CLI handles arguments.
- Invalid flags are now detected and make the execution stop (even in file/tree/branch lists)
- Adds various test cases to ensure the new functionality and old functionality both work

Addresses: #10

### CODE REVIEW

File: src/ReadSpeedCLI.cxx [hunk 1]
```diff
@@ -26,10 +26,10 @@ void ReadSpeed::PrintThroughput(const Result &r)
    std::cout << "Compressed throughput:\t\t" << r.fCompressedBytesRead / r.fRealTime / 1024 / 1024 << " MB/s\n";
 }
 
-Args ReadSpeed::ParseArgs(int argc, char **argv)
+Args ReadSpeed::ParseArgs(std::vector<std::string> args)
 {
    // Print help message and exit if "--help"
-   if (argc < 2 || (argc == 2 && (std::strcmp(argv[1], "--help") == 0 || std::strcmp(argv[1], "-h") == 0))) {
+   if (args.size() < 2 || (args.size() == 2 && (args[1].compare("--help") == 0 || args[1].compare("-h") == 0))) {
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx]: please use `str1 == str2` instead of `str1.compare(str2) == 0`, for readability

File: src/ReadSpeedCLI.cxx [hunk 2]
```diff
@@ -82,18 +82,30 @@ Args ReadSpeed::ParseArgs(int argc, char **argv)
...
    return Args{std::move(d), nThreads, branchState == EBranchState::kAll, /*fShouldRun=*/true};
+}
+
+Args ReadSpeed::ParseArgs(int argc, char **argv)
+{
+   std::vector<std::string> args;
+   args.reserve(argc);
+
+   for (int i = 0; i < argc; ++i) {
+      args.push_back(std::string(argv[i]));
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx]: less noisy and avoids a temporary: `args.emplace_back(argv[i])`

File: src/ReadSpeedCLI.cxx [hunk 3]
```diff
@@ -26,10 +26,10 @@ void ReadSpeed::PrintThroughput(const Result &r)
    std::cout << "Compressed throughput:\t\t" << r.fCompressedBytesRead / r.fRealTime / 1024 / 1024 << " MB/s\n";
 }
 
-Args ReadSpeed::ParseArgs(int argc, char **argv)
+Args ReadSpeed::ParseArgs(std::vector<std::string> args)
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx]: avoid a copy with `const std::vector<std::string> &`

File: src/ReadSpeedCLI.cxx [hunk 4]
```diff
@@ -48,8 +48,8 @@ Args ReadSpeed::ParseArgs(int argc, char **argv)
    const auto branchOptionsErrMsg =
       "Options --all-branches, --branches, and --branches-regex are mutually exclusive. You can use only one.\n";
 
-   for (int i = 1; i < argc; ++i) {
-      std::string arg(argv[i]);
+   for (size_t i = 1; i < args.size(); ++i) {
+      auto arg = args[i];
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx]: avoid a copy with `const auto &arg = args[i];`

File: test/tests.cpp [hunk 5]
```diff
@@ -139,4 +139,69 @@ TEST_CASE("CLI test")
                     "Number of parsed trees does not match number of provided trees.");
       CHECK_MESSAGE(outBranches == inBranches, "List of parsed trees does not match list of provided trees.");
    }
+   SUBCASE("Help arg")
+   {
+      const std::vector<std::string> allArgs{"root-readspeed", "--help"};
+
+      const auto parsedArgs = ParseArgs(allArgs);
+
+      CHECK_MESSAGE(!parsedArgs.fShouldRun, "Programme running when using help argument");
```
Comment (eguiraud) [inline test/tests.cpp]: Not really a problem, but in ROOT in general we use the US spelling "program"

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thanks @dkpoult ! Is there a test for the case mentioned in #10 , namely `--branches x --notanoption`? I did not see one.

Some more comments below.

**Comment by dkpoult** [comment]
> Is there a test for the case mentioned in #10 , namely `--branches x --notanoption`? I did not see one.

I've added one now, thanks for the catch!

File: src/ReadSpeedCLI.cxx [hunk 6]
```diff
@@ -64,33 +64,48 @@ Args ReadSpeed::ParseArgs(int argc, char **argv)
...
          if (branchState != EBranchState::kNone && branchState != EBranchState::kRegex) {
             std::cerr << branchOptionsErrMsg;
             return {};
          }
          branchState = EBranchState::kRegex;
          d.fUseRegex = true;
-      } else if (arg.compare("--threads") == 0) {
+      } else if (arg == "--threads") {
          argState = EArgState::kThreads;
+      } else if (arg.compare(0, 1, "-") == 0) {
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx]: more readable:

```suggestion
      } else if (arg[0] == '-') {
```

File: src/ReadSpeedCLI.cxx [hunk 7]
```diff
@@ -48,14 +48,14 @@ Args ReadSpeed::ParseArgs(int argc, char **argv)
    const auto branchOptionsErrMsg =
       "Options --all-branches, --branches, and --branches-regex are mutually exclusive. You can use only one.\n";
 
-   for (int i = 1; i < argc; ++i) {
-      std::string arg(argv[i]);
+   for (size_t i = 1; i < args.size(); ++i) {
+      const auto &arg = args[i];
```
Comment (eguiraud) [inline src/ReadSpeedCLI.cxx:52]: can this just be:
```suggestion
   for (const auto &arg : args) {
```
?

**Review by eguiraud — CHANGES_REQUESTED** [review]
Great, thanks! Just two last minor comments :)

File: src/ReadSpeedCLI.cxx (hunk 7 above)
Comment (dkpoult) [reply src/ReadSpeedCLI.cxx:52]: It needs to skip the first arg which would be "root-readspeed" or something similar so using a ranged-based loop is a bit more convoluted

Comment (eguiraud) [reply src/ReadSpeedCLI.cxx:52]: ah right, fair enough

**Review by eguiraud — APPROVED** [review]

**Pull request merged**